### PR TITLE
[Prim] Add instance_norm composite rule

### PR DESCRIPTION
**Pull request by rainyfly**
### PR types

Others
### PR changes

Others
### Describe

1.  Add instance_norm composite rule base on [PR51283](https://github.com/PaddlePaddle/Paddle/pull/51238)

Note:  No fp16 kernel is registered，and this op only supports NC*.

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py [hunk 1]
```diff
@@ -84,6 +86,535 @@ def _cal_mean_variance(x, epsilon, mean_shape):
...
+            places,
+            'float32',
+            [
+                [5e-5, 5e-5, 5e-5],  # cpu thresholds
+                [1e-4, 1e-4, 1e-4],
+            ],  # gpu thresholds
+            None,
+        ),
+        (
+            'test0_fp64',
```
Comment (cyber-pioneer) [inline python/paddle/fluid/tests/unittests/test_instance_norm_op.py:250]: add big shape case like [8,32,32,64]

File: python/paddle/incubate/autograd/composite_rules.py [hunk 2]
```diff
@@ -181,6 +181,37 @@ def layernorm_composite(x, scale, bias, epsilon, begin_norm_axis):
...
+    n, c, h, w = x.shape
+    axis = tuple(range(2, len(x.shape)))
+    mean_ = mean(x, axis=axis, keepdim=True)
+    difference = x - mean_
+    var_tmp1 = difference * difference
+    variance = mean(var_tmp1, axis=axis, keepdim=True)
+    var_tmp3 = variance + epsilon
+    sqrt_var = sqrt(var_tmp3)
+    saved_variance = 1 / sqrt_var
+    out = difference / sqrt_var
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py:196]: refer:
saved_variance = pow(var_tmp3, full(-0.5))
out =  difference * saved_variance

Comment (rainyfly) [reply python/paddle/incubate/autograd/composite_rules.py:196]: keep using divide for precision

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py (hunk 1 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_instance_norm_op.py:250]: done

File: python/paddle/incubate/autograd/composite_rules.py [hunk 3]
```diff
@@ -178,6 +178,37 @@ def layernorm_composite(x, scale, bias, epsilon, begin_norm_axis):
...
+    """
+
+    n, c, h, w = x.shape
+    axis = tuple(range(2, len(x.shape)))
+    mean_ = mean(x, axis=axis, keepdim=True)
+    difference = x - mean_
+    var_tmp1 = difference * difference
+    variance = mean(var_tmp1, axis=axis, keepdim=True)
+    var_tmp3 = variance + epsilon
+    sqrt_var = pow(var_tmp3, 0.5)
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py]: 0.5 -> full(0.5)

File: python/paddle/incubate/autograd/composite_rules.py [hunk 4]
```diff
@@ -178,6 +178,37 @@ def layernorm_composite(x, scale, bias, epsilon, begin_norm_axis):
...
 
+@REGISTER_COMPOSITE('instance_norm')
+def instancenorm_composite(x, scale, bias, epsilon):
+    """
+    define composite rule of op instance_norm
+    out = (x - mean(x)) / sqrt(var + epsilon))
+    var = mean((x-mean(x))^2)
+    """
+
+    n, c, h, w = x.shape
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py:188]: image
how about other data_format like nhwc?

File: python/paddle/incubate/autograd/composite_rules.py (hunk 3 above)
Comment (rainyfly) [reply python/paddle/incubate/autograd/composite_rules.py]: done

**Comment by rainyfly** [comment]
fp16 kernel is not supported by instance_norm op

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py [hunk 5]
```diff
@@ -84,6 +87,723 @@ def _cal_mean_variance(x, epsilon, mean_shape):
...
 
 
+def instance_norm_wrapper(x, weight=None, bias=None, esp=1e-05):
+    return paddle.nn.functional.instance_norm(
+        x, None, None, weight, bias, True, 0.9, esp
+    )
+
+
+class TestInstanceNormOp(OpTest):
+    def setUp(self):
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_instance_norm_op.py:96]: 这个测试cinn可以通过？ 如果没问题请加一下这个文件到cmake里面去

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py [hunk 6]
```diff
@@ -84,6 +87,723 @@ def _cal_mean_variance(x, epsilon, mean_shape):
...
+
+    def test_check_grad_cpu(self):
+        self.check_grad_with_place(
+            core.CPUPlace(),
+            ['X', 'Scale', 'Bias'],
+            'Y',
+            check_prim=True,
+        )
+
+    def test_check_output_gpu(self):
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_instance_norm_op.py]: 这个前向gpu测试跟cpu测试阈值都一样，把这两个测试合并了，统一调用check_output

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py [hunk 7]
```diff
@@ -84,6 +87,723 @@ def _cal_mean_variance(x, epsilon, mean_shape):
...
+        if core.is_compiled_with_cuda():
+            self.check_output_with_place(
+                core.CUDAPlace(0), atol=1e-05, check_prim=True
+            )
+
+    def test_check_grad_gpu(self):
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_instance_norm_op.py]: 这个反向也可以合并 调用check_grad

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py [hunk 8]
```diff
@@ -84,6 +87,723 @@ def _cal_mean_variance(x, epsilon, mean_shape):
...
+        self.rev_comp_rtol = 1e-13
+        self.rev_comp_atol = 1e-13
+        self.dtype = dtype
+
+    def test_check_output_cpu(self):
+        self.check_output_with_place(
+            core.CPUPlace(), atol=1e-14, check_prim=True
+        )
+
+    def test_check_output_gpu(self):
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_instance_norm_op.py]: 这两个测试函数可以删了直接用父类的就行吧

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py (hunk 5 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_instance_norm_op.py:96]: 不能，已经设置flag

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py (hunk 6 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_instance_norm_op.py]: done

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py (hunk 7 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_instance_norm_op.py]: done

File: python/paddle/fluid/tests/unittests/test_instance_norm_op.py (hunk 8 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_instance_norm_op.py]: done

**Review by cyber-pioneer — DISMISSED** [review]

**Review by cyber-pioneer — APPROVED** [review]

**Pull request merged**